File: strategy/generator/exit/mad.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from core.models.source import SourceType

from .base import Exit, ExitType


@dataclass(frozen=True)
class MadExit(Exit):
    type: ExitType = ExitType.Mad
    source: Parameter = StaticParameter(SourceType.CLOSE)
    period_fast: Parameter = StaticParameter(8.0)
    period_slow: Parameter = StaticParameter(23.0)

File: ta_lib/strategies/exit/src/mad.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;

pub struct MadExit {
    source: SourceType,
    period_fast: usize,
    period_slow: usize,
}

impl MadExit {
    pub fn new(source: SourceType, period_fast: f32, period_slow: f32) -> Self {
        Self {
            source,
            period_fast: period_fast as usize,
            period_slow: period_slow as usize,
        }
    }
}

impl Exit for MadExit {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period_fast, self.period_slow)
    }

    fn close(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let mad =
            data.source(self.source)
                .spread_pct(Smooth::SMA, self.period_fast, self.period_slow);

        (mad.cross_under(&ZERO), mad.cross_over(&ZERO))
    }
}
